### PR TITLE
coreutils: 9.3 -> 9.4

### PR DESCRIPTION
**Pull request by trofi**
Changes: https://lists.gnu.org/archive/html/info-gnu/2023-08/msg00007.html

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dasJ** [comment]
See also #252165 which also drops some patches. Can you drop them here as well?

**Comment by trofi** [comment]
> See also #252165 which also drops some patches. Can you drop them here as well?

AFAIU those were dropped around `9.0 -> 9.1` update (and a bit pater). `coreutils` does not apply any patches in `9.3`. I dropped `aarch64`-specific `sed '2i print "Skipping tail assert test"; exit 77' -i ./tests/tail-2/assert.sh` from there.

**Review by Ma27 — APPROVED** [review]

File: pkgs/tools/misc/coreutils/default.nix [hunk 1]
```diff
@@ -77,8 +77,6 @@ stdenv.mkDerivation rec {
       echo "int main() { return 77; }" > gnulib-tests/test-getlogin.c
     ''
   ])) + (optionalString stdenv.isAarch64 ''
-    sed '2i print "Skipping tail assert test"; exit 77' -i ./tests/tail-2/assert.sh
-
     # Sometimes fails: https://github.com/NixOS/nixpkgs/pull/143097#issuecomment-954462584
     sed '2i echo Skipping cut huge range test && exit 77' -i ./tests/misc/cut-huge-range.sh
```
Comment (Artturin) [inline pkgs/tools/misc/coreutils/default.nix:81]: file was moved https://github.com/coreutils/coreutils/blob/15925d0e5b80467c354fe27cb1205d6b6e3da11d/tests%2Fcut%2Fcut-huge-range.sh so build fails

```suggestion
    sed '2i echo Skipping cut huge range test && exit 77' -i ./tests/cut/cut-huge-range.sh
```

**Comment by Artturin** [comment]
Fedora has added only one new patch for 9.4 which is related to `--enable-systemd` which we don't do so not relevant for us
https://src.fedoraproject.org/rpms/coreutils/blob/rawhide/f/coreutils-9.4-systemd-coredump.patch

**Comment by minijackson** [comment]
I am getting some errors with coreutils 9.4 when using `stdenv.mkDerivation` with `src = ./some/local/dir`, during `unpackPhase`.

```bash
# With 9.3
/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin/cp -pr --reflink=auto -- /nix/store/...-vim vim
# Success!

# With 9.4
/nix/store/ifzwv2xqwdnv1gz87rxkizi67py5p3vj-coreutils-9.4/bin/cp -pr --reflink=auto -- /nix/store/...-vim vim
# Error:
# /nix/store/ifzwv2xqwdnv1gz87rxkizi67py5p3vj-coreutils-9.4/bin/cp: setting permissions for 'vim': Invalid argument
```

Can anyone confirm? I tried it on an ext4 filesystem.

**Comment by minijackson** [comment]
After a running both under strace, I can see this from 9.3:

```
mkdirat(AT_FDCWD, "vim", 0555)          = 0
newfstatat(AT_FDCWD, "vim", {st_mode=S_IFDIR|0555, st_size=4096, ...}, AT_SYMLINK_NOFOLLOW) = 0
openat(AT_FDCWD, "vim", O_RDONLY|O_NOFOLLOW|O_CLOEXEC|O_PATH) = 3
newfstatat(3, "", {st_mode=S_IFDIR|0555, st_size=4096, ...}, AT_EMPTY_PATH) = 0
chmod("/proc/self/fd/3", 040755)        = 0
close(3)                                = 0
```

And this from 9.4:

```
mkdirat(AT_FDCWD, "vim", 0555)          = 0
newfstatat(AT_FDCWD, "vim", {st_mode=S_IFDIR|0555, st_size=4096, ...}, AT_SYMLINK_NOFOLLOW) = 0
fchmodat2(AT_FDCWD, "vim", 040755, AT_SYMLINK_NOFOLLOW) = -1 EINVAL (Invalid argument)
```

It seems they changed the way they `chmod` directories. According to the `fchmodat` manpage, `EINVAL` is returned when "Invalid flag specified in flags", which here is `AT_SYMLINK_NOFOLLOW`.

Above we can see:

> AT_SYMLINK_NOFOLLOW: If pathname is a symbolic link, do not dereference it: instead operate on the link itself.  This flag is not currently implemented.

This makes me confused…

**Comment by minijackson** [comment]
Made a separate issue here #302046

**Pull request merged**